File: workspace/workflows/src/collection/begin_transition_to.rs
use crate::activity::{Activity, ActivityContext, Error};
use async_trait::async_trait;
use bosca_client::client::begin_transition::BeginTransitionInput;
use bosca_client::client::{Client, WorkflowJob};

pub struct CollectionBeginTransitionToActivity {
    id: String,
}

impl Default for CollectionBeginTransitionToActivity {
    fn default() -> Self {
        Self::new()
    }
}

impl CollectionBeginTransitionToActivity {
    pub fn new() -> CollectionBeginTransitionToActivity {
        CollectionBeginTransitionToActivity {
            id: "collection.begin.transition.to".to_string(),
        }
    }
}

#[async_trait]
impl Activity for CollectionBeginTransitionToActivity {
    fn id(&self) -> &String {
        &self.id
    }

    async fn execute(&self, client: &Client, _: &mut ActivityContext, job: &WorkflowJob) -> Result<(), Error> {
        let state = job
            .workflow_activity
            .configuration
            .get("state")
            .unwrap()
            .as_str()
            .unwrap()
            .to_string();
        let status_value = job.workflow_activity.configuration.get("status").unwrap();
        let status = if status_value.is_null() {
            "".to_string()
        } else {
            status_value.as_str().unwrap().to_string()
        };
        let metadata = &job.metadata.clone().unwrap();
        client
            .begin_transition(BeginTransitionInput {
                metadata_id: Some(metadata.id.to_owned()),
                version: Some(metadata.version.to_owned()),
                collection_id: None,
                state_id: state.to_owned(),
                status,
                supplementary_id: None,
                wait_for_completion: None
            })
            .await?;
        Ok(())
    }
}

File: workspace/runners/kotlin/activities/src/main/kotlin/io/bosca/workflow/metadata/BeginTransitionTo.kt
package io.bosca.workflow.metadata

import io.bosca.api.Client
import io.bosca.graphql.fragment.WorkflowJob
import io.bosca.graphql.type.ActivityInput
import io.bosca.util.toJsonElement
import io.bosca.util.toOptional
import io.bosca.workflow.Activity
import io.bosca.workflow.ActivityContext
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import java.text.DateFormat
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.util.Date

class BeginTransitionTo(client: Client) : Activity(client) {

    override val id: String = ID

    override suspend fun toActivityDefinition(): ActivityInput {
        return ActivityInput(
            id = id,
            name = "Begin Metadata Transition",
            description = "Begin a Metadata Transition",
            configuration = mapOf<String, Any>(
                "state" to "draft",
            ).toOptional(),
            inputs = emptyList(),
            outputs = emptyList(),
        )
    }

    override suspend fun execute(context: ActivityContext, job: WorkflowJob) {
        val configuration = job.workflowActivity.workflowActivity.configuration as Map<*, *>
        val requiredState = configuration["requiredState"] as? String
        if (requiredState != null && requiredState != job.metadata?.metadata?.workflow?.metadataWorkflow?.state) {
            return
        }
        val state = configuration["state"] as String
        val current = job.metadata?.metadata?.workflow?.metadataWorkflow
        if (state != current?.state && state != current?.pending) {
            val states = client.workflows.getStates()
            val published = states.first { it.id == "published" }
            var stateValid: ZonedDateTime? = null
            if (state == published.id) {
                var date: Date? = null
                job.metadata?.metadata?.let { metadata ->
                    metadata.attributes?.toJsonElement()?.let { attributes ->
                        val published = attributes.jsonObject.getValue("published").jsonPrimitive.longOrNull
                        if (published != null) {
                            date = Date(published)
                        } else {
                            val published = attributes.jsonObject.getValue("published").jsonPrimitive.contentOrNull
                            if (published != null) {
                                date = DateFormat.getDateTimeInstance().parse(published)
                            }
                        }
                    }
                }
                if (date != null && date.after(Date(System.currentTimeMillis()))) {
                    stateValid = ZonedDateTime.ofInstant(date.toInstant(), ZoneOffset.UTC)
                }
            }
            client.workflows.beginMetadataTransition(
                id = job.metadata?.metadata?.id ?: error("missing metadata"),
                version = job.metadata!!.metadata.version,
                state = state,
                status = "Begin Metadata Transition",
                stateValid = stateValid,
            )
        }
    }

    companion object {
        const val ID = "metadata.begin.transition.to"
    }
}